File: hibernate-criteria-builder/src/main/java/io/github/robertomike/hefesto/actions/wheres/WhereCustom.java
package io.github.robertomike.hefesto.actions.wheres;

import io.github.robertomike.hefesto.enums.WhereOperator;
import javax.persistence.criteria.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;
import java.util.Map;

@Getter
@Setter
@AllArgsConstructor
public class WhereCustom extends BaseWhere {
    private Custom custom;

    public WhereCustom(Custom custom, WhereOperator operator) {
        this.whereOperation = operator;
        this.custom = custom;
    }

    public interface Custom {
        Predicate call(CriteriaBuilder cb, CriteriaQuery<?> cr, Root<?> root, Map<String, Join<?, ?>> joins, Root<?> parentRoot);
    }
}

File: hibernate-criteria-builder/src/main/kotlin/io/github/robertomike/hefesto/actions/wheres/WhereCustom.kt
package io.github.robertomike.hefesto.actions.wheres

import io.github.robertomike.hefesto.enums.WhereOperator
import jakarta.persistence.criteria.*

class WhereCustom(val custom: Custom, whereOperator: WhereOperator = WhereOperator.AND) : BaseWhere() {
    init {
        this.whereOperation = whereOperator
    }

    fun interface Custom {
        fun call(
            cb: CriteriaBuilder,
            cr: CriteriaQuery<*>,
            root: Root<*>,
            joins: @JvmSuppressWildcards Map<String, Join<*, *>>,
            parentRoot: Root<*>
        ): Predicate
    }
}
